### ISSUE TITLE
@staticmethod with implicit object argument type is handled wrongly

### ISSUE THREAD

**Issue by navytux** [cython/cython#3174]
Hello up there. Please consider the following example:

---- 8< ---- `moda.pxd`
```pyx
cdef class MyClass:
    @staticmethod
    cdef static_func(x)
```

---- 8< ---- `moda.pyx`
```pyx
cdef class MyClass:
    @staticmethod
    cdef static_func(x):
        return x+1
```

---- 8< ---- `modb.pyx`
```pyx
from moda cimport MyClass

def test():
    x = MyClass.static_func(2)
    print(x)
```

When compiled and run it fails with:

```
$ python -c 'import modb; modb.test()'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "modb.pyx", line 4, in modb.test
    x = MyClass.static_func(2)
TypeError: Cannot convert int to moda.MyClass
```

If I add explicit `object` to `x` argument in `moda.pxd`:

```diff
--- a/moda.pxd
+++ b/moda.pxd
@@ -1,3 +1,3 @@
 cdef class MyClass:
     @staticmethod
-    cdef static_func(x)
+    cdef static_func(object x)

```

It works:

```
$ python -c 'import modb; modb.test()'
3
```

Thanks beforehand,
Kirill

/cc @robertwb

**Comment by navytux** [comment]
I forgot to mention that it fails for both current 0.29.x and master.
I added corresponding test in https://github.com/cython/cython/pull/3175.

**Comment by scoder** [comment]
My guess is that it fails because `@staticmethod` is not evaluated (correctly) in `.pxd` files. Or maybe it incorrectly overwrites the prior handling of the decorator in the implementation file somehow.

PR for that is welcome, but I can't say what needs to be done for that. Needs some analysis to find out where this should be fixed.

**Comment by navytux** [comment]
@scoder, thanks for feedback. My impression is that somehow the "pass" that automatically annotates first argument of a method of a `cdef class` with corresponding type overwrites `@staticmethod` effect somehow. But I did not looked into the code at all. For now I filed PR with test (#3175, it can be already merged), this issue, so that the bug is not forgotten, and in my codebase worked around the problem with explicitly using `object` type for the first argument.
